### PR TITLE
scripts: edtlib: Reduce code duplication in phandle/value list parsing

### PR DESCRIPTION
**Pull request by ulfalizer**
Most of the logic for initializing `clocks` and `pwms` is the same and
can be shared. Add an `EDT._simple_phandle_val_list()` helper function for
initializing `clocks`, `pwms`, and any other properties on the form

    <foo>s = <phandle value phandle value ...>

where the nodes pointed at by the phandles are expected to have a
`#<foo>-cells` property.

This should make it easier to add similar properties.

There's still some code duplication in the classes (`PWM`, `Clock`, etc.),
but also some differences, so I'm wondering if requiring a class for
each might be okay. Maybe some more class-related stuff could be
factored out later.

Piggyback some related cleanup:

 - Have `_phandle_val_list()` take the name of the `#foo-cells` property
   instead of a function for fetching the value. The pattern is always
   the same.

 - Have `_add_names()` just take the "foo" part of "foo-names". Same
   pattern everywhere.

 - Clean up some redundant comments for stuff that's already documented
   in docstrings

 - Fix error messages printed by `_named_cells()` ("GPIOS controller" ->
   "GPIO controller", etc.)

### CODE REVIEW

**Comment by pabigot** [comment]
On the principle that the third time you do the same thing you should generalize, and `gpio`, `pwm`, `clocks` makes three, this seems like a good idea as a code cleanup.

But there's no test that demonstrates it handles edge cases (like wrong number of values in a compound), or at least handles the three existing cases.  Do we not have a test suite for the new devicetree infrastructure?  Or is there a test suite that just didn't need to be modified?

**Comment by galak** [comment]
> But there's no test that demonstrates it handles edge cases (like wrong number of values in a compound), or at least handles the three existing cases. Do we not have a test suite for the new devicetree infrastructure? Or is there a test suite that just didn't need to be modified?

There are some testsuites:

scripts/dts/testdtlib.py
scripts/dts/testedtlib.py

File: scripts/dts/edtlib.py [hunk 1]
```diff
@@ -693,26 +687,55 @@ def _init_clocks(self):
...
+        #   'dev', 'controller', 'specifier', and 'name' fields initialized.
+        #   See the documentation for e.g. the PWM class.
+        #
+        # Returns a list of 'cls' instances.
 
-            pwm = PWM()
-            pwm.dev = self
-            pwm.controller = controller
-            pwm.specifier = self._named_cells(controller, spec, "pwms")
+        prop = self._node.props.get(name + "s")
```
Comment (galak) [inline scripts/dts/edtlib.py]: I don't think we should add the 's' here.  I think its a bit confusing on the caller and we should leave it to the caller to pass the name as it is in the .dts

**Review by galak — CHANGES_REQUESTED** [review]
I think `_simple_phandle_val_list` should not add the 's' and leave that to the caller.

**Comment by pabigot** [comment]
> There are some testsuites:
> 
> scripts/dts/testdtlib.py
> scripts/dts/testedtlib.py

Wow.  That's...really primitive.  And minimal.  And not very reassuring.

File: scripts/dts/edtlib.py (hunk 1 above)
Comment (ulfalizer) [reply scripts/dts/edtlib.py]: Had it like that at first, but it gets ugly in that you'd instead have to strip off a last character that may or may not be an `s` (and maybe add an ugly `assert` for it), so this way seemed nicer and cleaner. The ends-with-an-`s` requirement is enforced more directly.

Note that it's documented in the docstring.

**Comment by ulfalizer** [comment]
> Wow. That's...really primitive. And minimal.

It's very deliberate. Primitive and minimal and very simple to trace is what I strive for. :)

Coverage is probably better than you assume, especially for `dtlib`, which even verifies most of the error messages. The style just makes it compact.

The test suites just immediately `exit(1)` on the first failure, because we should never introduce any.

**Review by galak — APPROVED** [review]

**Pull request merged**